backend/src/import.cpp

```diff
@@ -0,0 +1,311 @@
+            buffer.resize(readPosition + actualRead + 1);
+            buffer.back() = '\0';
+            return actualRead < n;
+        }
+
+        void error(const char* msg) {
+            cerr << "\x1b[31mparse error:\x1b[0m " << msg << ":\n";
+            cerr << remaining().substr(0, 256) << endl;
+        }
+
+    private:
+        vector<char> buffer;
+        int position;
+};
+
+ParseResult parseWarning(ParseState& state) {
+    if (!state.remaining().starts_with("Warning:")) {
+        state.error("no warning message");
+        return ParseResult::FAIL;
+    }
+    auto end = state.remaining().find("\n\n");
+    if (end == string_view::npos) {
+        state.error("expected blank line after warning");
+        return ParseResult::FAIL;
+    }
+    state.advance(end);
+    return ParseResult::DONE;
+}
+
+int assignNodeIdx(Node& node) {
+    static auto idxMap = unordered_map<size_t, int>();
+    static auto hashFn = hash<string>();
+    auto hash = hashFn(node.data);
+    auto [idx, _] = idxMap.emplace(make_pair(hash, idxMap.size() + 1));
+    node.hash = to_string(hash);
+    node.idx = idx->second;
+    return node.idx;
+}
+
+ParseResult parseNodeBazel5(ParseState& state, Graph& graph) { /*
+    BLUE_NODE:NodeData{e247f4c}|
+    RED_NODE:NodeData{6566162}|GREEN_NODE:NodeData{aab8e2e}|BLUE_NODE:NodeData{e247f4c}|RED_NODE:NodeData{5cedeee}
+    RED_NODE:NodeData{5cedeee}|GREEN_NODE:NodeData{aab8e2e}
+    GREEN_NODE:NodeData{aab8e2e}|RED_NODE:NodeData{5cedeee}
+*/  auto source = 0;
+    while (true) {
+        auto eol = state.remaining().find("\n");
+        if (eol == string::npos) {
+            // line is partial
+            return ParseResult::MORE;
+        } else if (eol == 0) {
+            // skip empty lines
+            state.advance(1);
+            return ParseResult::DONE;
+        }
+        auto pipe = state.remaining().find("|");
+        auto end = pipe < eol ? pipe : eol;
+        auto data = string(state.remaining().substr(0, end));
+        auto colon = data.find(':');
+        if (colon == string::npos) {
+            state.error("missing colon in key");
+            return ParseResult::FAIL;
+        }
+        state.advance(end + 1);
+        string type = string(data.substr(0, colon));
+        auto idx = assignNodeIdx(graph.nodes.emplace_back(0, "", data, type));
+        if (source == 0) {
+            source = idx;
+        } else {
+            graph.edges.emplace_back(source, idx, 0);
+        }
+        if (end == eol) return ParseResult::DONE;
+    }
+}
+
+ParseResult parseNodeBazel6(ParseState& state, Graph& graph) { /*
+    BLUE_NODE:NodeData{e247f4c}
+    RED_NODE:NodeData{6566162}
+      Group 1:
+        BLUE_NODE:NodeData{e247f4c}
+        GREEN_NODE:NodeData{aab8e2e}
+      Group 2:
+        RED_NODE:NodeData{5cedeee}
+    RED_NODE:NodeData{5cedeee}
+      Group 1:
+        GREEN_NODE:NodeData{aab8e2e}
+    GREEN_NODE:NodeData{aab8e2e}
+      Group 1:
+        RED_NODE:NodeData{5cedeee}
```

Comment:
Example of Skyframe dump format from Bazel 6. One call to parseNodeBazel5 should parse, e.g.
```
   RED_NODE:NodeData{6566162}
      Group 1:
        BLUE_NODE:NodeData{e247f4c}
        GREEN_NODE:NodeData{aab8e2e}
      Group 2:
        RED_NODE:NodeData{5cedeee}
```
Which is equivalent to the following in Bazel 5:
```
RED_NODE:NodeData{6566162}|GREEN_NODE:NodeData{aab8e2e}|BLUE_NODE:NodeData{e247f4c}|RED_NODE:NodeData{5cedeee}
```
The parser should skip over any blank lines (of which there are several in the real dump format).